casr/src/bin/casr-ubsan.rs

```diff
@@ -0,0 +1,432 @@
+use casr::util;
+use libcasr::report::CrashReport;
+use libcasr::severity::Severity;
+use libcasr::stacktrace::{CrashLine, CrashLineExt};
+use libcasr::ubsan;
+use libcasr::ubsan::UbsanWarning;
+
+use anyhow::{bail, Context, Result};
+use clap::{
+    error::{ContextKind, ContextValue, ErrorKind},
+    Arg, ArgAction,
+};
+use log::{debug, info, warn};
+use rayon::iter::{IntoParallelRefIterator, ParallelIterator};
+use regex::Regex;
+use wait_timeout::ChildExt;
+use walkdir::WalkDir;
+
+use std::collections::HashSet;
+use std::env;
+use std::fs;
+use std::fs::OpenOptions;
+use std::io::{Read, Write};
+use std::path::{Path, PathBuf};
+use std::process::{Command, Stdio};
+use std::time::Duration;
+
+/// Extract ubsan warnings for specified input file
+///
+/// # Arguments
+///
+/// * `input` - input file path
+///
+/// * `argv` - target program argument vector
+///
+/// * `timeout` - target program timeout
+///
+/// # Returns value
+///
+/// Vector of extracted ubsan warnings with crashlines
+fn extract_warnings(
+    input: &PathBuf,
+    argv: &[&str],
+    timeout: u64,
+) -> Result<Vec<(UbsanWarning, CrashLine)>> {
+    // Get command line argv
+    let mut argv = argv.to_owned();
+    let arg: String;
+    let stdin = if let Some(index) = argv.iter().position(|&arg| arg.contains("@@")) {
+        arg = argv[index].replace("@@", input.to_str().unwrap());
+        argv[index] = &arg;
+        false
+    } else {
+        true
+    };
+    // Run program.
+    let mut cmd = Command::new(argv[0]);
+    cmd.stdout(Stdio::null()).stderr(Stdio::piped());
+    if stdin {
+        let Ok(file) = fs::File::open(input) else {
+            bail!("Can't open file {:?}", input);
+        };
+        cmd.stdin(file);
+    }
+    if argv.len() > 1 {
+        cmd.args(&argv[1..]);
+    }
+    debug!("Run: {:?}", cmd);
+
+    // Get stderr
+    let stderr: String =
+    // If timeout is specified, spawn and check timeout
+    // Else get output
+    if timeout != 0 {
+        let mut child = cmd
+            .spawn()
+            .with_context(|| "Failed to start command: {cmd:?}")?;
+        if child
+            .wait_timeout(Duration::from_secs(timeout))
+            .unwrap()
+            .is_none()
+        {
+            child.kill()?;
+            warn!("Timeout: {:?}", cmd);
+        }
+        let mut buf = vec![];
+        let _ = child.stderr.unwrap().read_to_end(&mut buf);
+        String::from_utf8_lossy(&buf).to_string()
+    } else {
+        let output = cmd
+            .output()
+            .with_context(|| "Failed to start command: {cmd:?}")?;
+        String::from_utf8_lossy(&output.stderr).to_string()
```

Comment:
this is already a String, no need to call to_string

Reply:
No, it's needed